PyNvCodec/src/PyNvCodec.cpp

```diff
@@ -220,6 +221,65 @@ bool PySurfaceDownloader::DownloadSingleSurface(shared_ptr<Surface> surface,
   return false;
 }
 
+PySurfaceToPtr::PySurfaceToPtr(){}
+
+bool PySurfaceToPtr::Execute(std::shared_ptr<Surface> surf, CUdeviceptr ptr) {
+  if (!surf) {
+    std::stringstream ss;
+    ss << __FUNCTION__;
+    ss << ": Source video frame has void CUDA device ptr.";
+    throw std::runtime_error(ss.str());
+  }
+
+  if (!ptr) {
+    std::stringstream ss;
+    ss << __FUNCTION__;
+    ss << ": Destination video frame has void CUDA device ptr.";
+    throw std::runtime_error(ss.str());
+  }
+
+  auto pPlane = surf->GetSurfacePlane(0U);
+  auto res = cudaMemcpy2D((void *) ptr, pPlane->Width(), (void *)pPlane->GpuMem(), pPlane->Pitch(),
+                          pPlane->Width(), pPlane->Height(), cudaMemcpyDeviceToDevice);
+  if (cudaSuccess != res) {
+    std::stringstream ss;
+    ss << __FUNCTION__;
+    ss << ": failed to copy surface data to tensor. CUDA error code: ";
+    ss << res;
+    throw std::runtime_error(ss.str());
+  }
+
+  return true;
+}
+
+PySurfaceFromPtr::PySurfaceFromPtr(uint32_t width, uint32_t height,
+                                   Pixel_Format format, uint32_t gpuID)
+    : height(height), width(width), gpuID(gpuID), outputFormat(format) {
+    surface.reset(Surface::Make(format, width, height,
+                                CudaResMgr::Instance().GetCtx(gpuID)));
+}
+
+shared_ptr<Surface> PySurfaceFromPtr::Execute(CUdeviceptr ptr) {
+  if (!ptr) {
+    return shared_ptr<Surface>(Surface::Make(outputFormat));
```

Comment:
should we return an exception:
```
  if (!ptr) {
    std::stringstream ss;
    ss << __FUNCTION__;
    ss << ":Source video frame has void CUDA device ptr.";
    throw std::runtime_error(ss.str());
  }
```

Reply:
It depends on what behavior you want to achieve. Basically there are 3 options:

- Return nullptr which will be mapped to `None` in Python.
- Throw exception.
- Return empty surface.

Last two are both okay-ish IMO. 
I generally prefer not to throw exceptions if it doesn't crash the program & return empty surface instead.